src/Ersatz/Solver/Z3.hs

```diff
@@ -0,0 +1,68 @@
+--------------------------------------------------------------------
+-- |
+-- Copyright :  © Edward Kmett 2010-2014, Johan Kiviniemi 2013
+-- License   :  BSD3
+-- Maintainer:  Edward Kmett <ekmett@gmail.com>
+-- Stability :  experimental
+-- Portability: non-portable
+--
+--------------------------------------------------------------------
+module Ersatz.Solver.Z3
+  ( z3
+  , z3Path
+  ) where
+
+import Data.ByteString.Builder
+import Control.Applicative
+import Control.Monad.IO.Class
+import Data.IntMap (IntMap)
+import Ersatz.Internal.Parser
+import Ersatz.Problem
+import Ersatz.Solution
+import Ersatz.Solver.Common
+import qualified Data.IntMap as IntMap
+import System.IO
+import System.Process (readProcessWithExitCode)
+
+-- | 'Solver' for 'SAT' problems that tries to invoke the @z3@ executable from the @PATH@
+z3 :: MonadIO m => Solver SAT m
+z3 = z3Path "z3"
+
+-- | 'Solver' for 'SAT' problems that tries to invoke a program that takes @z3@ compatible arguments.
+--
+-- The 'FilePath' refers to the path to the executable.
+z3Path :: MonadIO m => FilePath -> Solver SAT m
+z3Path path problem = liftIO $
+  withTempFiles ".cnf" "" $ \problemPath _ -> do
+    withFile problemPath WriteMode $ \fh ->
+      hPutBuilder fh (dimacs problem)
+
+    (_exit, out, _err) <-
+      readProcessWithExitCode path ["-dimacs", problemPath] []
+
+    let result = case lines out of
+                    "s SATISFIABLE":_   -> Satisfied
+                    "s UNSATISFIABLE":_ -> Unsatisfied
+                    _                   -> Unsolved
+
+    return (result, parseSolution out)
+
+parseSolution :: String -> IntMap Bool
+parseSolution input =
+  case runParser solution input of
+    s:_ -> s
+    _   -> IntMap.empty
+
+solution :: Parser Char (IntMap Bool)
+solution = do
+  _ <- string "s SATISFIABLE\nv "
```

Comment:
The use of `\n` here makes me a bit nervous, since I worry that this will choke on operating systems that use `\r\n` carriage returns (e.g., Windows). I wonder if we should err on the side of caution by introducing an [`endOfLine`](https://hackage.haskell.org/package/parsec-3.1.15.1/docs/Text-Parsec-Char.html#v:endOfLine) combinator, similar to the one in `parsec`, and use it here.